oio/directory/meta1.py

```diff
@@ -0,0 +1,117 @@
+
+    def _get_args_by_base(self, base):
+        return self.args_by_base.get(base, None)
+
+    def _apply_link_services(self, moved_ok, **kwargs):
+        for base in moved_ok:
+            peers = self._get_peers_by_base(base)
+            service_type = self._get_service_type_by_base(base)
+            args = self._get_args_by_base(base)
+            cid, seq = self.get_cid_and_seq(base)
+
+            try:
+                self.reference.force(
+                    service_type=service_type, cid=cid,  replace=True,
+                    services=dict(host=','.join(peers), type=service_type,
+                                  args=args, seq=seq))
+            except OioException as exc:
+                self.logger.warn(
+                    "Failed to link services for base %s (seq=%d): %s",
+                    cid, seq, exc)
+
+    @property
+    def reference(self):
+        if not self._reference:
+            self._reference = DirectoryClient(self.conf)
+        return self._reference
+
+    def move(self, src_service, dest_service, base_name, service_type,
+             **kwargs):
+        """
+        Move a `base` of `src_service` to `dest_service`
+        """
+        if service_type not in self.services_by_service_type.keys():
+            raise ValueError(
+                "service type must be %s"
+                % " or ".join(self.services_by_service_type.keys()))
+        cid, seq = self.get_cid_and_seq(base_name)
+
+        data = self.reference.list(cid=cid)
+        if dest_service not in \
+                self.services_by_service_type[service_type].keys():
+            raise ValueError(
+                "destination service must be a %s service" % service_type)
+
+        bases = dict()
+        for service in data['srv']:
+            if service['type'] != service_type:
+                continue
+            if seq is not None and seq != service['seq']:
+                continue
+            base = cid + "." + str(service['seq'])
+            raw_services = bases.get(base, None)
+            if raw_services is None:
+                raw_services = dict()
+                bases[base] = raw_services
+            host = service['host']
+            service.pop('host', None)
+            raw_services[host] = service
+
+        valid_bases = set()
+        for base, raw_services in bases.iteritems():
+            old_peers = raw_services.keys()
+            if src_service in old_peers and dest_service not in old_peers:
+                service = raw_services.pop(src_service)
+                raw_services[dest_service] = service
+                valid_bases.add(base)
+                self.moved_bases.add(base)
+                self.raw_services_by_base[base] = old_peers
+                self.services_by_base[base] = raw_services
+                self.service_type_by_base[base] = service_type
+                self.args_by_base[base] = service['args']
+
+        if not valid_bases:
```

Comment:
Couldn't you get more precise error reporting ?
"X or Y is wrong" is not super helpful as an error message, or maybe there's another way to know which one is wrong...